core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java

```diff
@@ -323,46 +324,51 @@ private String getCanonicalizedHeaderString(Map<String, List<String>> canonicali
         StringBuilder buffer = new StringBuilder();
 
         canonicalizedHeaders.forEach((headerName, headerValues) -> {
-            for (String headerValue : headerValues) {
-                appendCompactedString(buffer, headerName);
-                buffer.append(":");
-                if (headerValue != null) {
-                    appendCompactedString(buffer, headerValue);
-                }
-                buffer.append("\n");
-            }
+            buffer.append(headerName);
+            buffer.append(":");
+            buffer.append(String.join(",", trimAll(headerValues)));
```

Comment:
Again This is doing some extra memory allocations for the `String.join` that can be avoided. See https://github.com/aws/aws-sdk-java-v2/pull/2502/commits/632fea82eae1e3cafad6ab8bf7028d12eb45b31e